### PR TITLE
300409 apptail should persist tail position in file

### PR DESCRIPTION
**Pull request by Nino-K**

### CODE REVIEW

File: apptail/config.go [hunk 1]
```diff
@@ -12,10 +12,11 @@ type leakyBucketConfig struct {
...
 type Config struct {
-	MaxRecordSize     int    `json:"max_record_size"`
-	MaxLinesPerSecond int64  `json:"max_lines_per_second"`
-	MaxLinesBurst     uint16 `json:"max_lines_burst"`
-	FileSizeLimit     int64  `json:"read_limit"`
+	MaxRecordSize      int    `json:"max_record_size"`
+	MaxLinesPerSecond  int64  `json:"max_lines_per_second"`
+	MaxLinesBurst      uint16 `json:"max_lines_burst"`
+	FileSizeLimit      int64  `json:"read_limit"`
+	RetentionPerSecond int64  `json:"retention_per_second"`
```
Comment (Nino-K) [inline apptail/config.go]: Not too sure if the name ```RetentionPerSecond``` justify the operation??

Comment (Mook-as) [inline apptail/config.go]: Yeah, that name is a bit confusing.  Something along the lines of `PersistPositionIntervalSeconds` perhaps?  "Retention" is pretty vague, and it's an interval, not a frequency ("per second").

File: apptail/instance.go [hunk 2]
```diff
@@ -37,7 +38,7 @@ func (instance *Instance) Identifier() string {
 }
 
 // Tail begins tailing the files for this instance.
-func (instance *Instance) Tail() {
+func (instance *Instance) Tail(tracker storage.Tracker) {
```
Comment (Mook-as) [inline apptail/instance.go]: Should `Tracker` be an attribute inside of `Instance` (since each instance would belong to one tracker)?
(I could go either way on this, just wanted to get more discussion on it)

Comment (Nino-K) [inline apptail/instance.go]: That would change the singleton behaviour of the tracker though, since instance.Tail is being called per nats subscription in an anonymous way.
```
	natsclient.Subscribe("logyard."+uid+".newinstance", func(instance *apptail.Instance) {
		instance.Tail(tracker)
	})
```

File: apptail/config.go (hunk 1 above)
Comment (Nino-K) [inline apptail/config.go]: :+1:

File: apptail/instance.go (hunk 2 above)
Comment (Mook-as) [inline apptail/instance.go]: Right; I just tend to favour putting things that don't change over time to be composed, but that's more a personal quirk of disliking lots of function call arguments than some wisdom-based thing or anything like that.  Meh, I guess if it works as-is it's fine.

File: apptail/instance.go [hunk 3]
```diff
@@ -79,10 +95,16 @@ func (instance *Instance) tailFile(name, filename string, stopCh chan bool) {
...
 		Follow:      true,
-		Location:    &tail.SeekInfo{-limit, os.SEEK_END},
+		Location:    location,
 		ReOpen:      false,
 		Poll:        false,
 		RateLimiter: rateLimiter})
+
+	// IMPORTANT: this registration happens everytime app restarts
+	if shouldInitialize {
+		tracker.InitializeChildNode(instance.DockerId, filename, t.Location.Offset)
```
Comment (Mook-as) [inline apptail/instance.go]: Err, is `t.Location` the passed in `location` here, which has a `.Offset` of `-limit`?  (It's not using `t.Tell()` here?)

File: apptail/instance.go [hunk 4]
```diff
@@ -97,6 +119,12 @@ FORLOOP:
 				err = t.Wait()
 				break FORLOOP
 			}
+			currentOffset, err := t.Tell()
```
Comment (Mook-as) [inline apptail/instance.go]: The only way I see for `tail.TailFile()` to return an error is if the file doesn't exist; does trying to `t.Tell()` here ever make sense?  Also, does `t.Tell()` even return errors correctly? (It seems to return the uninitialized value of an `error`, and I have no idea what that is)

File: apptail/storage/gob_handler.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	return &FileStorage{file_path: path}
+
+}
+
+func (s *FileStorage) Write(data interface{}) {
+	m := new(bytes.Buffer)
+	enc := gob.NewEncoder(m)
+	err := enc.Encode(data)
+	if err != nil {
+		log.Error(err)
```
Comment (Mook-as) [inline apptail/storage/gob_handler.go]: Err, we log an error here, and then happily continue (writing out undefined bytes)? That doesn't sound like a good idea!

Also applies for other `log.Error()` calls in this function.

File: apptail/storage/gob_handler.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+	}
+	err = ioutil.WriteFile(s.file_path, m.Bytes(), FILE_MODE)
+	if err != nil {
+		log.Error(err)
+
+	}
+	// this extra step to make the file accessible by stackato user
+	if err = os.Chmod(s.file_path, FILE_MODE); err != nil {
```
Comment (Mook-as) [inline apptail/storage/gob_handler.go]: Does the last argument of `ioutil.WriteFile()` not work? Why is this needed?

File: apptail/storage/gob_handler.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		log.Error(err)
+
+	}
+}
+
+func (s *FileStorage) Load(e interface{}) {
+	n, err := ioutil.ReadFile(s.file_path)
+	if err != nil {
+		log.Error(err)
```
Comment (Mook-as) [inline apptail/storage/gob_handler.go]: Same thing as above about ignoring errors.  This function should probably return an optional `error` as well.

File: apptail/instance.go (hunk 3 above)
Comment (Nino-K) [inline apptail/instance.go]: We could use t.Tell() but we already have the current offset available to us, why do we need to use t.Tell()?

File: apptail/storage/tail_tracker.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+	RegisterInstance(instKey string)
+	InitializeChildNode(instKey string, childkey string, offSet int64)
+	Submit()
+	StartSubmissionTimer(retentionPeriod time.Duration)
+	IsInstanceRegistered(instKey string) bool
+	IsChildNodeInitialized(instKey string, childkey string) bool
+	GetFileCachedOffset(instkey string, fname string) int64
+}
+
+type BoxedInt64 struct{ V int64 }
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: Does `V` need to be exported? Why not `v`?  Actually, same with `BoxedInt64`.

File: apptail/storage/tail_tracker.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+var (
+	MinIOTicker = 5 * time.Second
+)
+
+func NewTracker(s Storage) Tracker {
+	return &tracker{
+		storage: s,
+		mux:     &sync.Mutex{},
+		Cached: &Tailer{
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: Needs `go fmt`

File: apptail/instance.go (hunk 3 above)
Comment (Mook-as) [inline apptail/instance.go]: Because the current offset you have says "negative 16 megabytes" or whatever; it's not the absolute position in the file at all.  (If you immediately restart `apptail` without reading anything first, it'll get serialized/deserialized and then try to see to negative 16MB from the start of the file, and fall over.)

File: apptail/storage/tail_tracker.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+	"time"
+)
+
+type Tracker interface {
+	Update(instKey string, childKey string, childVal int64)
+	LoadTailers()
+	Remove(key string)
+	RegisterInstance(instKey string)
+	InitializeChildNode(instKey string, childkey string, offSet int64)
+	Submit()
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: I think I'd prefer this to be called `Commit()` rather than `Submit()`, since that's what it actually does - it commits the changes to disk.

File: apptail/storage/tail_tracker.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+		storage: s,
+		mux:     &sync.Mutex{},
+		Cached: &Tailer{
+			Instances: make(map[string]TailNode),
+		},
+		timerStopChan: make(chan struct{}),
+	}
+}
+
+func (t *tracker) StartSubmissionTimer(retentionPeriod time.Duration) {
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: (Reminder that this will no longer be called `retentionPeriod`)

File: apptail/instance.go (hunk 4 above)
Comment (Nino-K) [inline apptail/instance.go]: >does trying to t.Tell() here ever make sense?

how elese can we get the current offset?

> Also, does t.Tell() even return errors correctly?

t.Tell() returns whatever tail.file.Seek would return while it's grabbing the offset.
Are  you suggesting to intercept the tailFile process if t.Tell() returns an error?

File: apptail/storage/tail_tracker.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+		}
+	}()
+}
+
+func (t *tracker) RegisterInstance(instKey string) {
+	t.mux.Lock()
+	if _, instance_exist := t.Cached.Instances[instKey]; !instance_exist {
+		t.Cached.Instances[instKey] = TailNode{}
+		t.formatMap("Current Status")
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: I wonder if we'll hit issues down the road from dumping out the status with the lock held.  But then, it's not like we can do this without the lock held…

File: apptail/storage/tail_tracker.go (hunk 8 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: Gob decoding and encoding required the fileds to be exported.

File: apptail/instance.go (hunk 4 above)
Comment (Mook-as) [inline apptail/instance.go]: > how elese can we get the current offset?

Err, yes, `t.Tell()` makes total sense here, I have no idea what I was smoking.

> Also, does t.Tell() even return errors correctly?

Ah, reading [`Tail.Tell()`](https://github.com/ActiveState/tail/blob/068b72961a6bc5b4a82cf4fc14ccc724c0cfa73a/tail.go#L122) again, it works by reusing `err`, so it does work fine.  I might be out of it when I made this comment :p

File: apptail/storage/tail_tracker.go (hunk 8 above)
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: Ah, okay, so we rely on gob to do the serialization; that means it (probably) won't attempt to do the read atomically either.  Which might be close enough for our purposes...  Still would prefer an `atomic.LoadInt64()` though.

File: apptail/storage/tail_tracker.go (hunk 11 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: :+1:

File: apptail/storage/tail_tracker.go (hunk 9 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: This code has already been throught gofmt || go fmt, not too sure why there is a wierd rendering :confused:

File: apptail/storage/tail_tracker.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	t.mux.Unlock()
+}
+
+// this is mainly used for testing since we are not exposing Cached via interface
+func (t *tracker) IsInstanceRegistered(instKey string) bool {
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: So, equivalent to
```go
func (t *tracker) IsInstanceRegistered(instKey string) (exist bool) {
    t.mux.Lock()
    defer t.mux.Unlock()
    _, exist = t.Cached.Instances[instKey]
}
```

File: apptail/storage/tail_tracker.go [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instKey]; instance_exist {
+		if _, childNode_exist := tailNode[childkey]; !childNode_exist {
+			tailNode[childkey] = &BoxedInt64{V: offSet}
+			t.Cached.Instances[instKey] = tailNode
+			t.formatMap("Current Status")
+		}
+	}
+	t.mux.Unlock()
+	runtime.Gosched()
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: Needs a comment on why `Gosched()` is needed. (What thread would this otherwise block? This doesn't seem to be a long-running loop…)

File: apptail/storage/tail_tracker.go [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+func (t *tracker) GetFileCachedOffset(instkey string, fname string) int64 {
+	var offset int64
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instkey]; instance_exist {
+		offset = atomic.LoadInt64(&tailNode[fname].V)
+	}
+	t.mux.Unlock()
+	runtime.Gosched()
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: Same as above; when using `Gosched()`, it probably deserves a comment on what it's unblocking.

File: apptail/storage/gob_handler.go (hunk 6 above)
Comment (Nino-K) [inline apptail/storage/gob_handler.go]: Not too sure, why it does not apply correct permissions while creating a file, it creates a file as root with only root rw.

File: apptail/storage/gob_handler.go (hunk 7 above)
Comment (Nino-K) [inline apptail/storage/gob_handler.go]: :+1:

File: apptail/storage/tail_tracker.go [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+func (t *tracker) GetFileCachedOffset(instkey string, fname string) int64 {
+	var offset int64
+	t.mux.Lock()
+	if tailNode, instance_exist := t.Cached.Instances[instkey]; instance_exist {
+		offset = atomic.LoadInt64(&tailNode[fname].V)
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: So why is an `offset` of `0` (I assume that's the default value) valid when the instance doesn't exist?  It sounds like you're not handling the error / passing it along, here.

File: apptail/storage/tail_tracker.go (hunk 10 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: :+1:

File: apptail/storage/tail_tracker.go [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	t.mux.Unlock()
+	runtime.Gosched()
+	return offset
+}
+
+func (t *tracker) Update(instKey string, childKey string, childVal int64) {
+	if tailNode, instance_exist := t.Cached.Instances[instKey]; instance_exist {
+		if _, childNode_exist := tailNode[childKey]; childNode_exist {
```
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: I guess the child wouldn't exist if that instance just got terminated... okay, that makes sense.

File: apptail/storage/tail_tracker.go (hunk 8 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: > atomic.LoadInt64()

we are doing that though.

Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: Err? I don't see that near [FileStorage.Write()](https://github.com/ActiveState/logyard-apps/pull/7/files#diff-0abd8b982345db71cd3b8ae245d05e6cR28).

File: apptail/storage/tail_tracker.go (hunk 16 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: This method is being called while the instance_key always exists, but good point error handeling wouldn't hurt here.

File: apptail/storage/tail_tracker.go (hunk 14 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: Initialized chiled noded runs per file basis(stderr, stdout, etc). So apptail loops throught each type of file and creates routines per file, the Gosched() allows the routine to run and yield to the next file asap.

File: apptail/storage/tail_tracker.go (hunk 15 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: Same with this method, it runs per file and Gosched yeilds to the next routine.

Comment (philwhln) [inline apptail/storage/tail_tracker.go]: Then this is relevant to the caller and not to this function, so should be outside of this function, not inside. The function itself is not doing anything blocking.

File: apptail/storage/tail_tracker.go [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+	t.mux.Lock()
+	t.storage.Load(&t.Cached)
+	t.formatMap("Loaded")
+	t.mux.Unlock()
+}
+
+func (t *tracker) Submit() {
+	t.mux.Lock()
+	t.formatMap("Storing")
+	t.storage.Write(t.Cached)
```
Comment (philwhln) [inline apptail/storage/tail_tracker.go]: I think this needs to be broken into 2 different mutexes to avoid blocking all of apptail on disk IO. First would be rendering what you want write to disk, which would use this mutex and block rest of apptail. Second would be for writing that to disk and uses a new mutex to ensure 2 callers don't write to disk at the same time.

File: apptail/storage/tail_tracker.go (hunk 16 above)
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: If we want to assume that `instance_key` always exists, then we shouldn't be using an `if`.  Just access it directly and let it abort when it fails.  At least, that seems to be the way go people want to handle things…

File: apptail/storage/tail_tracker.go (hunk 14 above)
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: Great! Now please put that as a comment in the code so we don't have to find this GitHub page when we look it again in three months :)

File: apptail/storage/tail_tracker.go (hunk 13 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: :+1:  updated!

File: apptail/storage/tail_tracker.go (hunk 16 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: Defensive programming ???!!!

File: apptail/storage/tail_tracker.go (hunk 18 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: Good point, updated!

File: apptail/storage/tail_tracker.go (hunk 16 above)
Comment (Mook-as) [inline apptail/storage/tail_tracker.go]: But this doesn't actually get us any benefit: things will mysteriously not work, since you don't handle the error condition.  I guess if you log it or something that's good enough? (The default panic behaviour includes logging, anyway...)

File: apptail/storage/tail_tracker.go (hunk 15 above)
Comment (Nino-K) [inline apptail/storage/tail_tracker.go]: :+1: correct! it should be called at the higher level.

File: apptail/instance.go (hunk 3 above)
Comment (Nino-K) [inline apptail/instance.go]: :+1: makes sense, updated!
